bigquery-writer/bigquery-writer.py

```diff
@@ -63,28 +85,69 @@ def write(self, listens):
             })
 
         body = { 'rows' : bq_data }
-        try:
-            t0 = time()
-            ret = self.bigquery.tabledata().insertAll(
-                projectId="listenbrainz",
-                datasetId="listenbrainz_test",
-                tableId="listen",
-                body=body).execute(num_retries=5)
-            self.time += time() - t0
-        except HttpError as e:
-            self.log.error("Submit to BigQuery failed: " + str(e))
-            self.log.error(json.dumps(body, indent=3))
+        while True:
+            try:
+                t0 = time()
+                ret = self.bigquery.tabledata().insertAll(
+                    projectId="listenbrainz",
+                    datasetId="listenbrainz_test",
+                    tableId="listen",
+                    body=body).execute(num_retries=5)
+                self.time += time() - t0
+                break
+
+            except HttpError as e:
+                self.log.error("Submit to BigQuery failed: %s. Sleeping 2 seconds." % str(e))
+            except Exception as e:
+                self.log.error("Unknown exception on submit to BigQuery failed: %s. Sleeping 2 seconds." % str(e))
+                if DUMP_JSON_WITH_ERRORS:
+                    self.log.error(json.dumps(body, indent=3))
+
+            sleep(2)
+
+
+        while True:
+            try:
+                self.channel.basic_ack(delivery_tag = method.delivery_tag)
+                break
+            except pika.exceptions.ConnectionClosed:
+                self.connect_to_rabbitmq()
+
+        self.redis.decr(UNIQUE_QUEUE_SIZE_KEY, count)
 
         # Clear the start time, since we've cleaned out the batch
         batch_start_time = 0
 
+        self.log.info("inserted %d listens." % count)
+
+        # collect and occasionally print some stats
+        self.inserts += count
+        if self.inserts >= REPORT_FREQUENCY:
+            self.total_inserts += self.inserts
+            if self.time > 0:
+                self.log.info("Inserted %d rows in %.1fs (%.2f listens/sec). Total %d rows." % \
+                    (self.inserts, self.time, count / self.time, self.total_inserts))
+            self.inserts = 0
+            self.time = 0
+
         return True
 
     def start(self):
+        self.log.info("biqquer-writer init")
+
+        if not hasattr(config, "REDIS_HOST"):
+            self.log.error("Redis service not defined. Sleeping 2 seconds and exiting.")
+            sleep(2)
+            return
+
+        if not hasattr(config, "RABBITMQ_HOST"):
+            self.log.error("RabbitMQ service not defined. Sleeping 2 seconds and exiting.")
+            sleep(2)
+            return
 
         # if we're not supposed to run, just sleep
         if not config.WRITE_TO_BIGQUERY:
-            sleep(1000)
+            sleep(66666)
```

Comment:
The actual value doesn't matter as long as it is large to prevent the container from restarting too often.